File: mobile/mobile/src/services/apiClient.ts
/**
 * API Client with Automatic Token Refresh
 *
 * Provides a configured axios instance with interceptors for:
 * - Automatic token attachment to requests
 * - Automatic token refresh on 401 responses
 * - Request retry after token refresh
 * - Logout on refresh failure
 *
 * Usage:
 * import { apiClient } from '../services/apiClient';
 * const response = await apiClient.get('/user/profile');
 */

import axios, { AxiosInstance, AxiosError, InternalAxiosRequestConfig } from 'axios';
import AsyncStorage from '@react-native-async-storage/async-storage';

/**
 * AsyncStorage keys for auth tokens
 * Must match keys used in AuthContext
 */
const ACCESS_TOKEN_KEY = '@AIParenting:accessToken';
const REFRESH_TOKEN_KEY = '@AIParenting:refreshToken';
const USER_DATA_KEY = '@AIParenting:userData';

/**
 * API base URL from environment variable
 * Defaults to localhost for development
 */
const API_URL = process.env.API_URL || 'http://localhost:3000';

/**
 * Flag to prevent multiple simultaneous refresh attempts
 * When one request triggers a refresh, others wait for it to complete
 */
let isRefreshing = false;

/**
 * Queue of failed requests waiting for token refresh
 * These requests will be retried after refresh succeeds
 */
let failedQueue: Array<{
  resolve: (value?: any) => void;
  reject: (error?: any) => void;
}> = [];

/**
 * Process the queue of failed requests
 * Called after token refresh succeeds or fails
 *
 * @param error - Error to reject all requests with (if refresh failed)
 * @param token - New access token to use for retrying requests (if refresh succeeded)
 */
const processQueue = (error: any = null, token: string | null = null) => {
  failedQueue.forEach((promise) => {
    if (error) {
      promise.reject(error);
    } else {
      promise.resolve(token);
    }
  });

  // Clear the queue
  failedQueue = [];
};

/**
 * Create axios instance with default configuration
 */
export const apiClient: AxiosInstance = axios.create({
  baseURL: API_URL,
  timeout: 30000, // 30 second timeout for requests
  headers: {
    'Content-Type': 'application/json',
  },
});

/**
 * Request Interceptor
 *
 * Automatically attaches access token to all requests.
 * Token is retrieved from AsyncStorage before each request.
 */
apiClient.interceptors.request.use(
  async (config: InternalAxiosRequestConfig) => {
    // Get access token from storage
    const accessToken = await AsyncStorage.getItem(ACCESS_TOKEN_KEY);

    // If token exists, add it to Authorization header
    if (accessToken) {
      config.headers.Authorization = `Bearer ${accessToken}`;
    }

    return config;
  },
  (error) => {
    // Request setup failed - return error
    return Promise.reject(error);
  }
);

/**
 * Response Interceptor
 *
 * Automatically handles 401 Unauthorized responses by:
 * 1. Attempting to refresh the access token
 * 2. Retrying the original request with new token
 * 3. Logging out user if refresh fails
 */
apiClient.interceptors.response.use(
  // Success response - pass through unchanged
  (response) => response,

  // Error response - check if token refresh is needed
  async (error: AxiosError) => {
    const originalRequest: any = error.config;

    // Check if error is 401 Unauthorized
    if (error.response?.status === 401 && originalRequest && !originalRequest._retry) {
      // Mark request as retried to prevent infinite loop
      originalRequest._retry = true;

      // If already refreshing, queue this request to be retried after refresh completes
      if (isRefreshing) {
        return new Promise((resolve, reject) => {
          failedQueue.push({ resolve, reject });
        })
          .then((token) => {
            // Refresh succeeded - retry original request with new token
            originalRequest.headers.Authorization = `Bearer ${token}`;
            return apiClient(originalRequest);
          })
          .catch((err) => {
            // Refresh failed - reject this request
            return Promise.reject(err);
          });
      }

      // Start refresh process
      isRefreshing = true;

      try {
        // Get refresh token from storage
        const refreshToken = await AsyncStorage.getItem(REFRESH_TOKEN_KEY);

        if (!refreshToken) {
          // No refresh token available - logout user
          throw new Error('No refresh token available');
        }

        // Call refresh endpoint to get new access token
        const response = await axios.post(`${API_URL}/auth/refresh`, {
          refreshToken,
        });

        const { accessToken: newAccessToken } = response.data;

        // Store new access token in AsyncStorage
        await AsyncStorage.setItem(ACCESS_TOKEN_KEY, newAccessToken);

        // Update Authorization header for original request
        originalRequest.headers.Authorization = `Bearer ${newAccessToken}`;

        // Process queued requests with new token
        processQueue(null, newAccessToken);

        // Retry original request with new token
        return apiClient(originalRequest);
      } catch (refreshError) {
        // Token refresh failed - logout user
        console.error('Token refresh failed:', refreshError);

        // Reject all queued requests
        processQueue(refreshError, null);

        // Clear all auth data from storage
        await AsyncStorage.multiRemove([
          ACCESS_TOKEN_KEY,
          REFRESH_TOKEN_KEY,
          USER_DATA_KEY,
        ]);

        // Reject original request
        // App-level navigation logic will detect missing tokens and show login screen
        return Promise.reject(refreshError);
      } finally {
        // Reset refreshing flag
        isRefreshing = false;
      }
    }

    // Not a 401 error or already retried - return original error
    return Promise.reject(error);
  }
);

/**
 * Helper function to set auth token after login
 * Used by AuthContext after successful login/registration
 *
 * @param token - Access token to set for subsequent requests
 */
export const setAuthToken = (token: string | null) => {
  if (token) {
    apiClient.defaults.headers.common['Authorization'] = `Bearer ${token}`;
  } else {
    delete apiClient.defaults.headers.common['Authorization'];
  }
};

/**
 * Helper function to clear auth token on logout
 * Used by AuthContext on logout
 */
export const clearAuthToken = () => {
  delete apiClient.defaults.headers.common['Authorization'];
};

export default apiClient;

File: mobile/lib/shared/services/api_client.dart
/// HTTP API client using Dio with automatic token refresh interceptor.
/// Handles all communication with the backend API including authentication,
/// error handling, and request/response logging.
library;

import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';
import 'api_config.dart';
import 'secure_storage_service.dart';

/// Main API client for making HTTP requests to the backend.
/// Includes automatic JWT token injection and refresh on 401 errors.
class ApiClient {
  final Dio _dio;
  final SecureStorageService _secureStorage;

  /// Constructor accepts Dio instance and SecureStorageService for dependency injection
  ApiClient(this._dio, this._secureStorage) {
    _setupInterceptors();
  }

  /// Factory constructor that creates instance with default Dio configuration
  factory ApiClient.create(SecureStorageService secureStorage) {
    // Create Dio instance with base configuration
    final dio = Dio(
      BaseOptions(
        baseUrl: ApiConfig.baseUrl,
        connectTimeout: Duration(seconds: ApiConfig.connectTimeout),
        receiveTimeout: Duration(seconds: ApiConfig.receiveTimeout),
        sendTimeout: Duration(seconds: ApiConfig.sendTimeout),
        headers: ApiConfig.defaultHeaders,
        validateStatus: (status) {
          // Accept all status codes for custom error handling
          return status != null && status < 500;
        },
      ),
    );

    return ApiClient(dio, secureStorage);
  }

  /// Sets up Dio interceptors for logging, auth, error handling, and retry logic
  void _setupInterceptors() {
    // Add logging interceptor in debug mode only
    if (kDebugMode) {
      _dio.interceptors.add(
        LogInterceptor(
          requestBody: true,
          responseBody: true,
          requestHeader: true,
          responseHeader: false,
          error: true,
          logPrint: (object) => debugPrint(object.toString()),
        ),
      );
    }

    // Add retry interceptor for transient errors (before auth interceptor)
    _dio.interceptors.add(_createRetryInterceptor());

    // Add authentication interceptor
    _dio.interceptors.add(
      InterceptorsWrapper(
        // Inject access token before each request
        onRequest: (options, handler) async {
          // Skip token injection for login/register/refresh endpoints
          final skipAuthPaths = [
            ApiConfig.loginEndpoint,
            ApiConfig.registerEndpoint,
            ApiConfig.refreshTokenEndpoint,
          ];

          if (!skipAuthPaths.contains(options.path)) {
            // Get access token from secure storage
            final accessToken = await _secureStorage.getAccessToken();
            if (accessToken != null) {
              // Add Bearer token to Authorization header
              options.headers['Authorization'] = 'Bearer $accessToken';
            }
          }

          // Continue with request
          handler.next(options);
        },

        // Handle response errors
        onError: (error, handler) async {
          // Handle 401 Unauthorized - token expired
          if (error.response?.statusCode == 401) {
            // Attempt to refresh token
            final refreshed = await _refreshToken();
            if (refreshed) {
              // Retry original request with new token
              try {
                final response = await _retry(error.requestOptions);
                return handler.resolve(response);
              } catch (e) {
                // If retry fails, pass error to caller
                return handler.next(error);
              }
            } else {
              // Refresh failed, user needs to login again
              await _secureStorage.clearAuthSession();
              // TODO: Navigate to login screen
              return handler.next(error);
            }
          }

          // Handle 429 Too Many Requests - usage limit reached
          if (error.response?.statusCode == 429) {
            debugPrint('Usage limit reached: ${error.response?.data}');
            // Let the caller handle this (show paywall)
          }

          // Pass error to caller
          handler.next(error);
        },
      ),
    );
  }

  /// Creates retry interceptor with exponential backoff for transient errors
  /// Automatically retries failed requests for server errors (500, 502, 503) and timeouts
  InterceptorsWrapper _createRetryInterceptor() {
    // Maximum number of retry attempts
    const maxRetries = 3;

    return InterceptorsWrapper(
      onError: (error, handler) async {
        // Check if error is retryable
        if (!_shouldRetry(error)) {
          // Not retryable, pass to next interceptor
          return handler.next(error);
        }

        // Get current retry count from request options extra data
        final retries = error.requestOptions.extra['retries'] as int? ?? 0;

        // Check if we've exceeded max retries
        if (retries >= maxRetries) {
          debugPrint('Max retries ($maxRetries) exceeded for ${error.requestOptions.path}');
          return handler.next(error);
        }

        // Calculate delay using exponential backoff: 1s, 2s, 4s
        final delayMs = (1000 * (1 << retries)); // 2^retries seconds in milliseconds
        debugPrint('Retrying request (attempt ${retries + 1}/$maxRetries) after ${delayMs}ms delay...');

        // Wait before retrying
        await Future.delayed(Duration(milliseconds: delayMs));

        // Increment retry count
        error.requestOptions.extra['retries'] = retries + 1;

        // Retry the request
        try {
          final response = await _dio.fetch(error.requestOptions);
          return handler.resolve(response);
        } on DioException catch (e) {
          // Retry failed, pass error to next interceptor
          return handler.next(e);
        }
      },
    );
  }

  /// Determines if an error should be retried
  /// Returns true for server errors (500, 502, 503) and timeout errors
  /// Returns false for client errors (4xx) and other errors
  bool _shouldRetry(DioException error) {
    // Don't retry if response is null (likely connection error, let it fail immediately)
    if (error.response == null) {
      // Retry on timeout errors
      if (error.type == DioExceptionType.connectionTimeout ||
          error.type == DioExceptionType.receiveTimeout ||
          error.type == DioExceptionType.sendTimeout) {
        return true;
      }
      // Don't retry other connection errors (no internet, etc.)
      return false;
    }

    final statusCode = error.response!.statusCode;

    // Retry on server errors (500, 502, 503)
    if (statusCode == 500 || statusCode == 502 || statusCode == 503) {
      return true;
    }

    // Don't retry on client errors (4xx) - these indicate problems with the request
    // Note: 401 is handled separately by auth interceptor
    if (statusCode != null && statusCode >= 400 && statusCode < 500) {
      return false;
    }

    // Don't retry other errors
    return false;
  }

  /// Attempts to refresh the access token using the refresh token
  /// Returns true if successful, false otherwise
  Future<bool> _refreshToken() async {
    try {
      // Get refresh token from storage
      final refreshToken = await _secureStorage.getRefreshToken();
      if (refreshToken == null) {
        debugPrint('No refresh token available');
        return false;
      }

      // Call refresh endpoint
      final response = await _dio.post(
        ApiConfig.refreshTokenEndpoint,
        data: {'refreshToken': refreshToken},
      );

      // Check if refresh was successful
      if (response.statusCode == 200) {
        // Extract new tokens from response
        final newAccessToken = response.data['accessToken'] as String;
        final newRefreshToken = response.data['refreshToken'] as String;

        // Save new tokens to secure storage
        await _secureStorage.saveAccessToken(newAccessToken);
        await _secureStorage.saveRefreshToken(newRefreshToken);

        debugPrint('Token refreshed successfully');
        return true;
      }

      return false;
    } catch (e) {
      debugPrint('Token refresh failed: $e');
      return false;
    }
  }

  /// Retries a failed request with new access token
  Future<Response> _retry(RequestOptions requestOptions) async {
    // Get new access token
    final newAccessToken = await _secureStorage.getAccessToken();

    // Create new options with updated token
    final options = Options(
      method: requestOptions.method,
      headers: {
        ...requestOptions.headers,
        'Authorization': 'Bearer $newAccessToken',
      },
    );

    // Retry the request
    return _dio.request(
      requestOptions.path,
      data: requestOptions.data,
      queryParameters: requestOptions.queryParameters,
      options: options,
    );
  }

  /// GET request helper method
  /// Throws DioException on network errors
  Future<Response> get(
    String path, {
    Map<String, dynamic>? queryParameters,
    Options? options,
  }) async {
    return await _dio.get(
      path,
      queryParameters: queryParameters,
      options: options,
    );
  }

  /// POST request helper method
  /// Throws DioException on network errors
  /// Optional onSendProgress callback for tracking upload progress
  Future<Response> post(
    String path, {
    dynamic data,
    Map<String, dynamic>? queryParameters,
    Options? options,
    void Function(int, int)? onSendProgress,
  }) async {
    return await _dio.post(
      path,
      data: data,
      queryParameters: queryParameters,
      options: options,
      onSendProgress: onSendProgress,
    );
  }

  /// PUT request helper method
  /// Throws DioException on network errors
  Future<Response> put(
    String path, {
    dynamic data,
    Map<String, dynamic>? queryParameters,
    Options? options,
  }) async {
    return await _dio.put(
      path,
      data: data,
      queryParameters: queryParameters,
      options: options,
    );
  }

  /// PATCH request helper method
  /// Throws DioException on network errors
  Future<Response> patch(
    String path, {
    dynamic data,
    Map<String, dynamic>? queryParameters,
    Options? options,
  }) async {
    return await _dio.patch(
      path,
      data: data,
      queryParameters: queryParameters,
      options: options,
    );
  }

  /// DELETE request helper method
  /// Throws DioException on network errors
  Future<Response> delete(
    String path, {
    dynamic data,
    Map<String, dynamic>? queryParameters,
    Options? options,
  }) async {
    return await _dio.delete(
      path,
      data: data,
      queryParameters: queryParameters,
      options: options,
    );
  }

  /// Upload file using multipart form data
  /// Used for photo uploads
  Future<Response> uploadFile(
    String path,
    String filePath,
    String fieldName, {
    Map<String, dynamic>? additionalData,
  }) async {
    // Create multipart form data
    final formData = FormData.fromMap({
      fieldName: await MultipartFile.fromFile(filePath),
      if (additionalData != null) ...additionalData,
    });

    return await post(path, data: formData);
  }

  /// Downloads file from URL
  /// Saves to specified file path with progress callback
  Future<Response> downloadFile(
    String url,
    String savePath, {
    void Function(int received, int total)? onProgress,
  }) async {
    return await _dio.download(
      url,
      savePath,
      onReceiveProgress: onProgress,
    );
  }
}
